### PR TITLE
build images for all supported pg versions #137

### PR DESCRIPTION
**Pull request by tavallaie**
change about build images for all supported pg versions #137

### CODE REVIEW

File: .github/workflows/extension_ci.yml [hunk 1]
```diff
@@ -151,25 +152,24 @@ jobs:
           trunk publish pgmq --version ${pgmq_ver} --description "A lightweight message queue. Like AWS SQS and RSMQ, on Postgres." --homepage "https://github.com/tembo-io/pgmq" --repository "https://github.com/tembo-io/pgmq" --license "PostgreSQL" --category featured --category orchestration
   build_and_push:
     name: Build and push images
-    # needs:
-    #   - publish
-    runs-on:
-      - self-hosted
-      - dind
-      - large-8x8
```
Comment (ChuckHend) [inline .github/workflows/extension_ci.yml:159]: pgrx can take quite a bit of resources. I think we'll want to use the self hosted runners for this.

File: .github/workflows/extension_ci.yml [hunk 2]
```diff
@@ -190,26 +190,24 @@ jobs:
           password: ${{ secrets.QUAY_PASSWORD_TEMBO }}
 
       - name: Build and push -- Commit
-        # push commit build when not a release
         if: github.event_name != 'release'
         uses: docker/build-push-action@v5
         with:
-          file: ./images/pgmq-pg/Dockerfile
+          file: ./images/pgmq-${{ matrix.pg_version }}/Dockerfile
```
Comment (ChuckHend) [inline .github/workflows/extension_ci.yml:196]: Does this line need to change? There is no dir `./images/pgmq-pg16` for example and we should be able to use a single dockerfile, right?

File: images/pgmq-pg/Dockerfile [hunk 3]
```diff
@@ -29,18 +30,18 @@ RUN $HOME/.cargo/bin/cargo pgrx install --pg-config=$(which pg_config)
...
-COPY --from=builder /usr/lib/postgresql/15/lib /usr/lib/postgresql/15/lib
+COPY --from=builder /usr/share/postgresql/${PG_VERSION#pg}/extension /usr/share/postgresql/${PG_VERSION#pg}/extension
+COPY --from=builder /usr/lib/postgresql/${PG_VERSION#pg}/lib /usr/lib/postgresql/${PG_VERSION#pg}/lib
 
 RUN apt-get update \
-    && apt-get install -y ca-certificates
+	&& apt-get install -y ca-certificates
 
-COPY ./images/pgmq-pg/postgresql.conf /usr/share/postgresql/15/postgresql.conf.sample
+COPY ./images/pgmq-${PG_VERSION}/postgresql.conf /usr/share/postgresql/${PG_VERSION#pg}/postgresql.conf.sample
```
Comment (ChuckHend) [inline images/pgmq-pg/Dockerfile:44]: Maybe?

```suggestion
COPY ./images/pgmq-pg/postgresql.conf /usr/share/postgresql/${PG_VERSION#pg}/postgresql.conf.sample
```

**Comment by ChuckHend** [comment]
Thank you for picking this up!

File: .github/workflows/extension_ci.yml (hunk 1 above)
Comment (ChuckHend) [reply .github/workflows/extension_ci.yml:159]: once we are completely moved off pgrx then this should no longer be a concern

Comment (tavallaie) [reply .github/workflows/extension_ci.yml:159]: I am not familiar with your build system yet so some times I miss configurations.

**Comment by ChuckHend** [comment]
I need to figure out why [docker login](https://github.com/tembo-io/pgmq/actions/runs/9244770049/job/25430450379?pr=233) is not working properly on this PR.

**Comment by v0idpwn** [comment]
Probably because it's running from a fork without the secrets? Sec...

**Comment by v0idpwn** [comment]
Yep, I created #235 and it passes this part but fails for other reasons.

**Comment by v0idpwn** [comment]
@tavallaie for testing, I think you can comment out the login step. Then it will try to build and fail at pushing. Once it fails at pushing only, we can merge :)

After merge, we can probably disable auth/push for PRs, these should run only on main branch.

**Comment by tavallaie** [comment]
it seems logical to me, contributor should not have such access someone can extract secret from main repositories :)

**Comment by tavallaie** [comment]
@v0idpwn I think I messed up somewhere in action and I need help.

**Comment by v0idpwn** [comment]
@tavallaie you added the input but you need to set it as an environment variable when building the docker image

**Comment by tavallaie** [comment]
it try to get postgresql-server-dev-15.7 but the last version I can find is [15.6-0+deb12u1](https://packages.debian.org/search?keywords=postgresql-server-dev-15) so how can prevent that?

File: .github/workflows/extension_ci.yml [hunk 4]
```diff
@@ -151,25 +152,27 @@ jobs:
...
   build_and_push:
     name: Build and push images
-    # needs:
-    #   - publish
     runs-on:
       - self-hosted
       - dind
       - large-8x8
+    strategy:
+      matrix:
```
Comment (v0idpwn) [inline .github/workflows/extension_ci.yml:160]:
```suggestion
      fail-fast: false
      matrix:
```
I'm interested in how it behaves with other versions

**Comment by v0idpwn** [comment]
@tavallaie The image is overriding PG_VERSION: https://github.com/docker-library/postgres/blob/8a0b96710d917d1c3b32a5fe5b66687ad83827da/15/bookworm/Dockerfile#L93

Maybe use a different name?

File: images/pgmq-pg/Dockerfile [hunk 5]
```diff
@@ -1,9 +1,10 @@
-FROM postgres:15.4-bookworm as builder
+# syntax=docker/dockerfile:1.4
 
-ARG PGMQ_VER=0.33.1
+ARG PG_DOCKER_VERSION
+FROM postgres:${PG_DOCKER_VERSION}-bookworm as builder
```
Comment (v0idpwn) [inline images/pgmq-pg/Dockerfile:4]: Need to redeclare the argument to keep using it afterwards, see https://docs.docker.com/reference/dockerfile/#understand-how-arg-and-from-interact

File: images/pgmq-pg/Dockerfile [hunk 6]
```diff
@@ -29,18 +31,19 @@ RUN $HOME/.cargo/bin/cargo pgrx install --pg-config=$(which pg_config)
...
 # install pg_partman
 RUN git clone https://github.com/pgpartman/pg_partman.git && \
-    cd pg_partman && \
-    git checkout v4.7.4 && make && make install
+	cd pg_partman && \
+	git checkout v4.7.4 && make && make install
 
-FROM postgres:15.4-bookworm
+FROM postgres:${PG_DOCKER_VERSION}-bookworm
+ARG PG_DOCKER_VERSION
```
Comment (v0idpwn) [inline images/pgmq-pg/Dockerfile:38]:
```suggestion
ARG PG_DOCKER_VERSION=$PG_DOCKER_VERSION
FROM postgres:${PG_DOCKER_VERSION}-bookworm
ARG PG_DOCKER_VERSION
```

File: images/pgmq-pg/Dockerfile (hunk 5 above)
Comment (tavallaie) [reply images/pgmq-pg/Dockerfile:4]: Before this pull request, I was just a simple Python developer. Now, thanks to your patience and guidance, I've leveled up! Thank you!

File: images/pgmq-pg/Dockerfile [hunk 7]
```diff
@@ -29,18 +31,20 @@ RUN $HOME/.cargo/bin/cargo pgrx install --pg-config=$(which pg_config)
 
 # install pg_partman
 RUN git clone https://github.com/pgpartman/pg_partman.git && \
-    cd pg_partman && \
-    git checkout v4.7.4 && make && make install
+	cd pg_partman && \
+	git checkout v4.7.4 && make && make install
 
-FROM postgres:15.4-bookworm
+ARG PG_DOCKER_VERSION=$PG_DOCKER_VERSION
```
Comment (v0idpwn) [inline images/pgmq-pg/Dockerfile:37]:
```suggestion
ARG PG_DOCKER_VERSION=${PG_DOCKER_VERSION}
```
not sure but seems to work here (?)

File: images/pgmq-pg/Dockerfile [hunk 8]
```diff
@@ -12,15 +14,15 @@ RUN apt-get update \
...
+	postgresql-server-dev-${PG_DOCKER_VERSION}
 
 # Install Rust dependencies
 RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
 RUN $HOME/.cargo/bin/rustup default stable
 
 # init pgrx
 RUN $HOME/.cargo/bin/cargo install cargo-pgrx --version=0.11.0 --locked
-RUN $HOME/.cargo/bin/cargo pgrx init --pg15 $(which pg_config)
+RUN $HOME/.cargo/bin/cargo pgrx init --pg$${PG_DOCKER_VERSION} $(which pg_config)
```
Comment (v0idpwn) [inline images/pgmq-pg/Dockerfile:25]:
```suggestion
RUN $HOME/.cargo/bin/cargo pgrx init --pg${PG_DOCKER_VERSION} $(which pg_config)
```

**Comment by tavallaie** [comment]
should we pass it as two different ARGs? based on [moby project](https://github.com/moby/moby/issues/37345)
it should be fixed now but some how we don't see expected result.

**Comment by v0idpwn** [comment]
I'm not sure. In my machine it works, but I'm using a newer version (26.1.3), the CI is running 20.10.23

File: images/pgmq-pg/Dockerfile [hunk 9]
```diff
@@ -29,18 +31,20 @@ RUN $HOME/.cargo/bin/cargo pgrx install --pg-config=$(which pg_config)
 
 # install pg_partman
 RUN git clone https://github.com/pgpartman/pg_partman.git && \
-    cd pg_partman && \
-    git checkout v4.7.4 && make && make install
+	cd pg_partman && \
+	git checkout v4.7.4 && make && make install
 
-FROM postgres:15.4-bookworm
+ARG PG_DOCKER_VERSION=${PG_DOCKER_VERSION}
```
[the remark this answers is not in the document]
Comment (ChuckHend) [reply images/pgmq-pg/Dockerfile:37]: Or was this was we tried previously?

**Comment by v0idpwn** [comment]
@tavallaie I think we can hold this PR for a while. We will do a transition from pgrx to pgrxs, which will simplify our build process and allow this dockerfile to be a single stage (what would magically solve the problem)

**Comment by tavallaie** [comment]
Greate, should I close it or leave it open?

**Comment by v0idpwn** [comment]
Let's leave it open for now :)

**Comment by v0idpwn** [comment]
Closing this in favor of #237. I went for a "Dockerfile template" which then has variables replaced. I'd be really happy to change that if we find a better approach (maybe taking 2 build args?). Still, this one will have too many conflicts after merge, as there are many CI changes and changes in this image, so probably starting clear will be better

**Pull request closed without merging**